### DISCUSSION TITLE
Gaussian Splatting on ROCm

### DISCUSSION THREAD

**Discussion by expenses** [ROCm/ROCm#2581] [Show and tell]
Hey guys, I've recently been interested in Gaussian Splattings and have ported the code from https://github.com/graphdeco-inria/gaussian-splatting to ROCm. The issue for this is over at https://github.com/graphdeco-inria/gaussian-splatting/issues/212#issuecomment-1767690373 and you can try out the code with:

```
git clone https://github.com/expenses/gaussian-splatting/
cd gaussian-splatting
git submodule init
git submodule update --recursive
```
and then by following the instructions in the readme.

I'd love to have more people trying this out and testing it as I'm pretty new to ROCm and I'm not sure if any performance problems are as result of my exotic setup or something else.

**Comment by fluidnumericsJoe** [comment]
Long times in CopyHostToDevice are not surprising with an external GPU . Do you know how much data is being copied? It'd be good to estimate achieved bandwidth and compare with theoretical bandwidth for your setup.

You could also build and run this [memcpy microbenchmark](https://github.com/FluidNumerics/gpu-programming/tree/main/samples/c%2B%2B/memcpy) on you system; you will need the HIP compiler to build it. The microbenchmark will tell you the achieved bandwidth for a range of array sizes.

If I have time, I'll try to test your demo on a bare metal system not using an external GPU.

**Comment by lji72** [comment]
@fluidnumerics-joe @expenses  Has this issue solved? and now I‘m try to run 3D GS on AMD Mi250 GPU.

**Comment by fluidnumericsJoe** [comment]
@lji72 - I haven't had time to benchmark this application on our MI210's. It's still in the queue but paying customers keep us busy :) 
Let us know if you give it a go

**Reply by liuji72** [reply]
ok, The project can be complied, but it has been blocked due to something unknown.

**Reply by fluidnumericsJoe** [reply]
Blocked how? It doesn't run at all? It throws an error you don't understand? Computer catches fire?

**Reply by lji72** [reply]
The program was stuck after several iters' training. I suppose it's caused by the Cooperative Groups API. These are some HIP code converted by HIPify.

template<int C>
__global__ void preprocessCUDA(
        int P, int D, int M,
        const float3* means,
        const int* radii,
        const float* shs,
        const bool* clamped,
        const glm::vec3* scales,
        const glm::vec4* rotations,
        const float scale_modifier,
        const float* proj,
        const glm::vec3* campos,
        const float3* dL_dmean2D,
        glm::vec3* dL_dmeans,
        float* dL_dcolor,
        float* dL_dcov3D,
        float* dL_dsh,
        glm::vec3* dL_dscale,
        glm::vec4* dL_drot)
{
        auto idx = cg::this_grid().thread_rank();
        if (idx >= P || !(radii[idx] > 0))
                return;

        float3 m = means[idx];

        // Taking care of gradients from the screenspace points


and call with this code:
hipLaunchKernelGGL((    preprocessCUDA<NUM_CHANNELS>) ,  dim3((P + 255) / 256), dim3(256) , 0, 0,


I'm not sure whether it is right or not.

**Reply by lji72** [reply]
Can you give me some guide about how to use  Cooperative Groups API in RoCM

**Reply by fluidnumericsJoe** [reply]
It seems it's currently not supported, but is going to be worked on. You could upvote the issue posted below to help raise priority.

https://github.com/ROCm/HIP-CPU/issues/60

**Comment by ZJLi2013** [comment]
hi, @expenses , tried to play with your repo as following:

```sh
git clone -b rocm https://github.com/expenses/gaussian-splatting.git
git submodule init && git submodule update --recursive 
sudo docker run --rm -it  --device=/dev/kfd --device=/dev/dri --group-add video --ipc=host --shm-size 16G --security-opt seccomp=unconfined --security-opt apparmor=unconfined -e HIP_VISIBLE_DEVICES=6 -v /home/zhengjli/aigc:/workspace -w /workspace  rocm/pytorch:rocm6.1.3_ubuntu22.04_py3.10_pytorch_release-2.1.2

# install conda 
wget https://repo.anaconda.com/miniconda/Miniconda3-py310_23.11.0-2-Linux-x86_64.sh 
#==> install path:  /root/conda 
bash Miniconda3* 
export PATH="$PATH:/root/conda/bin"
conda init bash
source /root/.bashrc

# install gs env
cd gaussian-splatting
conda env create --file environment.yml
``` 

but it gives:

```yml

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [13 lines of output]
      Traceback (most recent call last):
        File "<string>", line 36, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/workspace/gaussian-splatting/submodules/diff-gaussian-rasterization/setup.py", line 29, in <module>
          extra_compile_args={"nvcc": ["-I" + os.path.join(os.path.dirname(os.path.abspath(__file__)), "third_party/glm/")]})
        File "/opt/conda/envs/gaussian_splatting/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 983, in CUDAExtension
          library_dirs += library_paths(cuda=True)
        File "/opt/conda/envs/gaussian_splatting/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 1098, in library_paths
          if (not os.path.exists(_join_cuda_home(lib_dir)) and
        File "/opt/conda/envs/gaussian_splatting/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 2125, in _join_cuda_home
          raise EnvironmentError('CUDA_HOME environment variable is not set. '
      OSError: CUDA_HOME environment variable is not set. Please set it to your CUDA install root.
      No ROCm runtime is found, using ROCM_HOME='/opt/rocm'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

``` 

looks in your `rocm` branch, there is still explicitly depends on `cudatoolkit`?

thanks for direct me the right way to rebuild
David

**Comment by ZJLi2013** [comment]
ok,  I directly install the submodules, without conda, it works now

```sh
conda deactivate 
pip install submodules/simple-knn 
pip install submodules/diff-gaussian-rasterization

# train test
python3 train.py -s xxx
``` 
Thanks a lot

**Comment by ZJLi2013** [comment]
just confirmed: it does can run through on MI308. 

```yml

root@26a2670c9cb2:/workspace/gaussian-splatting# python3 train.py --source_path /workspace/dataset/tandt/train   --model_path  /workspace/output/
Optimizing /workspace/output/
Output folder: /workspace/output/ [01/08 10:18:24]
Reading camera 301/301 [01/08 10:18:26]
Converting point3d.bin to .ply, will happen only the first time you open the scene. [01/08 10:18:26]
Loading Training Cameras [01/08 10:18:27]
Loading Test Cameras [01/08 10:18:30]
Number of points at initialisation :  182686 [01/08 10:18:30]
Training progress:  23%|█████████████████▌                                                         | 7000/30000 [06:20<20:33, 18.65it/s, Loss=0.0875616]
[ITER 7000] Evaluating train: L1 0.07123592644929887 PSNR 19.83501319885254 [01/08 10:24:55]

[ITER 7000] Saving Gaussians [01/08 10:24:55]
Training progress: 100%|██████████████████████████████████████████████████████████████████████████| 30000/30000 [28:38<00:00, 17.46it/s, Loss=0.0624178]

[ITER 30000] Evaluating train: L1 0.03926312401890755 PSNR 24.267305755615237 [01/08 10:47:10]

[ITER 30000] Saving Gaussians [01/08 10:47:10]
```


* gt as  
![0000](https://github.com/user-attachments/assets/575804a2-e43f-4c28-9945-512cfa8b9f4a)

* rendered as 
![00000](https://github.com/user-attachments/assets/fbf5f36e-bff8-4ff2-abcb-90a5d68348c1)
